### PR TITLE
rgw/sync: add copying of mdlog entries in fullsync()

### PR DESCRIPTION
**Pull request by thotz**
In metadata incremental sync, RGWCloneMetaLogCoroutine copies mdlog entries from the primary zone, so that it can serve those entries to other zones in the case of failover (and it's promoted to the 'metadata master'), these entries are not copied in metadata full sync. so a new zone that syncs all of its metadata from full sync won't have any of the mdlog entries. if that new zone is promoted to the metadata master, it won't be able to serve these entries to other zones hat are behind in incremental sync. so those zones will think they're caught up, but may be missing lots of metadata

Fixes: https://tracker.ceph.com/issues/55770

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/driver/rados/rgw_sync.cc [hunk 1]
```diff
@@ -1647,6 +1647,95 @@ class RGWMetaSyncShardCR : public RGWCoroutine {
         collect_children();
       } while (omapkeys->more && can_adjust_marker);
 
+      mdlog_marker = sync_marker.marker;
+      marker = max_marker = sync_marker.marker;
+      do {
+#define INCREMENTAL_MAX_ENTRIES 100
```
Comment (mattbenjamin) [inline src/rgw/driver/rados/rgw_sync.cc:1653]: let's not copy-paste any #define magic numbers code, even though some older code does this.

File: src/rgw/driver/rados/rgw_sync.cc [hunk 2]
```diff
@@ -1647,6 +1647,95 @@ class RGWMetaSyncShardCR : public RGWCoroutine {
         collect_children();
       } while (omapkeys->more && can_adjust_marker);
 
+      mdlog_marker = sync_marker.marker;
```
Comment (mattbenjamin) [inline src/rgw/driver/rados/rgw_sync.cc:1650]: what do the different "markers" (all set to the value of sync_marker.marker) intend to do?  a comment would be helpful here

**Review by cbodley — COMMENTED** [review]
we'll need to clone the logs from each period separately. it's not clear to me how this interacts with the loop over periods in RGWMetaSyncCR: https://github.com/ceph/ceph/blob/84296aa/src/rgw/driver/rados/rgw_sync.cc#L2011-L2013

**Comment by thotz** [comment]
jenkins test make check arm64

**Comment by thotz** [comment]
jenkins test api

**Comment by thotz** [comment]
I have tested the following scenario:

- Created cluster with master zone `us-east` and add `us-west` to it
- Created the user "a1" and "bkt" via s3. Upload the file into it
- Added Zone `us-north` into, sync status was successful. Was able to fetch bkt/file via a1 user credentials and us-north endpoint
- Now killed server in `us-east` and change master zone to `us-north`
- Added zone `us-south` into it, sync status throw an error  `failed to fetch datalog info` but I was able to fetch bkt/file via a1 user credentials using `us-south` endpoint.

**Comment by thotz** [comment]
> I have tested the following scenario:
> 
>     * Created cluster with master zone `us-east` and add `us-west` to it
> 
>     * Created the user "a1" and "bkt" via s3. Upload the file into it
> 
>     * Added Zone `us-north` into, sync status was successful. Was able to fetch bkt/file via a1 user credentials and us-north endpoint
> 
>     * Now killed server in `us-east` and change master zone to `us-north`
> 
>     * Added zone `us-south` into it, sync status throw an error  `failed to fetch datalog info` but I was able to fetch bkt/file via a1 user credentials using `us-south` endpoint.

Above scenario works even without this PR

**Comment by thotz** [comment]
Closing this PR since no longer issue is seen in multicluster enviroment

**Comment by smanjara** [comment]
apart from testing, the PR looks ok to me. @cbodley is there more to be done here?

**Comment by cbodley** [comment]
failover may happen multiple times, generating a new period each time. each period has its own separate mdlog. incremental metadata sync will loop over periods in order and process the mdlogs for each until it's caught up with the current period. so the metadata master zone is expected to serve those mdlogs for all periods that its peers are still referencing

this pr calls RGWCloneMetaLogCoroutine for the current period, but doesn't try to clone the logs for any earlier periods. i think we still need to cover that part, and build test cases that exercise it

**Comment by smanjara** [comment]
> failover may happen multiple times, generating a new period each time. each period has its own separate mdlog. incremental metadata sync will loop over periods in order and process the mdlogs for each until it's caught up with the current period. so the metadata master zone is expected to serve those mdlogs for all periods that its peers are still referencing
> 
> this pr calls RGWCloneMetaLogCoroutine for the current period, but doesn't try to clone the logs for any earlier periods. i think we still need to cover that part, and build test cases that exercise it

agreed, thanks! i'll follow up with an update in https://github.com/ceph/ceph/pull/70829

**Pull request closed without merging**